### PR TITLE
Review

### PR DESCRIPTION
**Pull request by stefan-wullems**

### CODE REVIEW

File: src/QueryManager.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    this._ongoingQueries = {}
+
+    this.subscriptions = {}
+    this.statuses = {}
+  }
+
+  generateQueryRef (queries: Queries) {
+    const terms = this._extractTerms(queries)
+
+    let queryRef = JSON.stringify(terms)
```
Comment (stefan-wullems) [inline src/QueryManager.ts]: const instead of let for the queryRef

File: src/QueryManager.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  Subscriptions,
+  QueryRefs,
+  Statuses,
+  QueryCacheOptions,
+} from './types'
+import { Transform, RecordOperation, RecordIdentity } from '@orbit/data'
+import { identityIsEqual } from './helpers'
+
+export class QueryManager<E extends { [key: string]: any } = any>  {
+  _extensions: E
```
Comment (hongaar) [inline src/QueryManager.ts]: make explicit with `private _extensions: E` ?

File: src/QueryManager.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    let queryRef = JSON.stringify(terms)
+
+    if (this.subscriptions[queryRef]) return queryRef
+
+    this.subscriptions[queryRef] = { listeners: 0, terms }
+    return queryRef
+  }
+
+  _extractTerms (queries: Queries): Term[] {
```
Comment (hongaar) [inline src/QueryManager.ts]: make explicit with `private _extractTerms ...` 
http://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers

**Comment by hongaar** [comment]
nit: remove and ignore `yarn-error.log`

**Comment by hongaar** [comment]
would be nice to have some docs with e.g. introduction, example use case with code, api interface.

File: src/QueryManager.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    this.subscriptions = {}
+    this.statuses = {}
+  }
+
+  generateQueryRef (queries: Queries) {
+    const terms = this._extractTerms(queries)
+
+    let queryRef = JSON.stringify(terms)
+
+    if (this.subscriptions[queryRef]) return queryRef
```
Comment (hongaar) [inline src/QueryManager.ts]: I favour single return points in functions when possible (and I don't really like a single line if statements), e.g.:

```
if (!this.subscriptions[queryRef]) {
  this.subscriptions[queryRef] = { listeners: 0, terms }
}

return queryRef
```

File: src/QueryManager.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    const queries: Promise<RecordObject>[] = terms
+      .map(({ key, expression }) =>
+        new Promise((resolve, reject) => {
+          this._store.query(expression)
+            .then(record => resolve({ [key]: record }))
+            .catch(reject)
+        })
+      )
+
+    // The statuses[statusRef] object can be deleted before this gets called if client unsubscribes
```
Comment (Michiel87) [inline src/QueryManager.ts]: Would be nice if the status is not deleted until everything is resolved, regardless of amount of listeners.

File: src/QueryManager.ts [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  _generateStatusRef (queryRef: string) {
+    let i = 1
+    while (true) {
+      const statusRef = queryRef + `_${i}`
+      if (!this.statuses[statusRef]) return statusRef
+      else i++
+    }
+  }
+
+  async subscribeToFetch (queryRef: string, listener: () => void) {
```
Comment (Michiel87) [inline src/QueryManager.ts]: Where is this method going to be used? You should avoid async methods in components/hooks. 

Im assuming some of the functionality of this method but I think you should perform the queries strictly in the manager itself and subscribe only to its status and result.

File: src/QueryManager.ts [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    if (this._ongoingQueries[queryRef]) this._ongoingQueries[queryRef].listeners++
+    else throw new Error('There is no fetch going on with this queryRef. You possibly forgot to make a query before subscribing')
+
+    try {
+      await Promise.all(this._ongoingQueries[queryRef].queries);
+      listener();
```
Comment (Michiel87) [inline src/QueryManager.ts]: inconsistent semicolons

File: src/QueryManager.ts [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    if (this._ongoingQueries[queryRef].listeners === 0) {
+      delete this._ongoingQueries[queryRef]
+    }
+  }
+
+  unsubscribeFromCache (queryRef: string) {
+    this.subscriptions[queryRef].listeners--
+
+    if (this.subscriptions[queryRef].listeners === 0) {
+      delete this.subscriptions[queryRef]
```
Comment (Michiel87) [inline src/QueryManager.ts]: I dont think it makes a difference now but its not pure.

**Pull request closed without merging**